File: lib/screens/task/task_stats_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:fl_chart/fl_chart.dart';
import '../../l10n/app_localizations.dart';
import '../../theme/app_theme.dart';
import '../../data/task/task.dart';
import '../../providers/task_provider.dart';
import '../../utils/datetime/date_utils.dart';
import '../../utils/stats/stats_utils.dart';
import '../../widgets/common/week_navigator.dart';

class TaskStatsScreen extends StatefulWidget {
  const TaskStatsScreen({super.key});

  @override
  State<TaskStatsScreen> createState() => _TaskStatsScreenState();
}

class _TaskStatsScreenState extends State<TaskStatsScreen> {
  // This tracks which week we are currently viewing
  DateTime _focusedDate = DateTime.now();

  @override
  Widget build(BuildContext context) {
    final tasks = context.watch<TaskProvider>().tasks;
    final l10n = AppLocalizations.of(context)!;

    // Calculate week bounds
    final startOfWeek = AppDateUtils.startOfWeekMonday(_focusedDate);
    final endOfWeek = startOfWeek.add(
      const Duration(days: 6, hours: 23, minutes: 59, seconds: 59),
    );

    // Check if we are viewing the current real-world week
    final bool isCurrentRealWeek = AppDateUtils.isWithinInclusive(
      DateTime.now(),
      startOfWeek,
      endOfWeek,
    );

    return Scaffold(
      body: tasks.isEmpty
          ? Center(child: Text(l10n.taskStatsEmpty))
          : ListView(
              padding: const EdgeInsets.all(16),
              children: [
                // --- WEEK NAVIGATOR ---
                WeekNavigator(
                  focusedDate: _focusedDate,
                  onShiftWeek: _changeWeek,
                ),
                const SizedBox(height: 16),

                // TASK VELOCITY CHART
                Text(
                  l10n.taskStatsVelocity,
                  style: const TextStyle(
                    fontSize: 18,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                const SizedBox(height: 4),
                Row(
                  children: [
                    const Icon(
                      Icons.circle,
                      size: 12,
                      color: AppTheme.warningColor,
                    ),
                    const SizedBox(width: 4),
                    Text(
                      l10n.taskStatsCreated,
                      style: const TextStyle(fontSize: 12, color: Colors.grey),
                    ),
                    const SizedBox(width: 16),
                    const Icon(
                      Icons.circle,
                      size: 12,
                      color: AppTheme.successColor,
                    ),
                    const SizedBox(width: 4),
                    Text(
                      l10n.taskStatsCompleted,
                      style: const TextStyle(fontSize: 12, color: Colors.grey),
                    ),
                  ],
                ),
                const SizedBox(height: 16),
                SizedBox(
                  height: 200,
                  child: _buildVelocityChart(
                    context,
                    tasks,
                    startOfWeek,
                    endOfWeek,
                    isCurrentRealWeek,
                  ),
                ),
                const SizedBox(height: 40),

                // PRODUCTIVITY BAR CHART
                Text(
                  l10n.taskStatsProductivity,
                  style: const TextStyle(
                    fontSize: 18,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                const SizedBox(height: 4),
                Text(
                  l10n.taskStatsCompletedDaily,
                  style: const TextStyle(fontSize: 12, color: Colors.grey),
                ),
                const SizedBox(height: 16),
                SizedBox(
                  height: 200,
                  child: _buildProductivityChart(
                    context,
                    tasks,
                    startOfWeek,
                    endOfWeek,
                    isCurrentRealWeek,
                  ),
                ),
                const SizedBox(height: 40),

                // CATEGORY BREAKDOWN (ALL TIME)
                _buildCategoryBreakdownCard(context, tasks),
                const SizedBox(height: 40),
              ],
            ),
    );
  }

  // --- CHART BUILDERS ---
  Widget _buildVelocityChart(
    BuildContext context,
    List<Task> tasks,
    DateTime startOfWeek,
    DateTime endOfWeek,
    bool isCurrentRealWeek,
  ) {
    final velocityData = _getVelocityForWeek(tasks, startOfWeek, endOfWeek);
    final todayIndex = DateTime.now().weekday - 1;
    final l10n = AppLocalizations.of(context)!;
    final weekdays = [
      l10n.commonWeekdayMon,
      l10n.commonWeekdayTue,
      l10n.commonWeekdayWed,
      l10n.commonWeekdayThu,
      l10n.commonWeekdayFri,
      l10n.commonWeekdaySat,
      l10n.commonWeekdaySun,
    ];

    final maxY = StatsUtils.maxValue(
      velocityData.expand((day) => day),
      minimum: 5,
    );

    return BarChart(
      BarChartData(
        alignment: BarChartAlignment.spaceAround,
        maxY: maxY + 1,
        barTouchData: const BarTouchData(enabled: false),
        titlesData: FlTitlesData(
          bottomTitles: AxisTitles(
            sideTitles: SideTitles(
              showTitles: true,
              getTitlesWidget: (double value, TitleMeta meta) {
                final int dayIndex = value.toInt();
                // Only highlight the day if we are actually viewing the current week!
                final bool isToday =
                    isCurrentRealWeek && (dayIndex == todayIndex);

                return Padding(
                  padding: const EdgeInsets.only(top: 8.0),
                  child: Text(
                    weekdays[dayIndex],
                    style: TextStyle(
                      fontSize: 12,
                      fontWeight: isToday ? FontWeight.bold : FontWeight.normal,
                      color: isToday
                          ? AppTheme.taskColor
                          : Theme.of(context).colorScheme.onSurfaceVariant,
                    ),
                  ),
                );
              },
            ),
          ),
          leftTitles: const AxisTitles(),
          topTitles: const AxisTitles(),
          rightTitles: const AxisTitles(),
        ),
        gridData: const FlGridData(show: false),
        borderData: FlBorderData(show: false),
        barGroups: List.generate(7, (index) {
          return BarChartGroupData(
            x: index,
            barsSpace: 4,
            barRods: [
              BarChartRodData(
                toY: velocityData[index][0].toDouble(),
                color: AppTheme.warningColor,
                width: 10,
                borderRadius: BorderRadius.circular(2),
              ),
              BarChartRodData(
                toY: velocityData[index][1].toDouble(),
                color: AppTheme.successColor,
                width: 10,
                borderRadius: BorderRadius.circular(2),
              ),
            ],
          );
        }),
      ),
    );
  }

  Widget _buildProductivityChart(
    BuildContext context,
    List<Task> tasks,
    DateTime startOfWeek,
    DateTime endOfWeek,
    bool isCurrentRealWeek,
  ) {
    final dailyCounts = _getTasksCompletedForWeek(
      tasks,
      startOfWeek,
      endOfWeek,
    );
    final todayIndex = DateTime.now().weekday - 1;
    final l10n = AppLocalizations.of(context)!;
    final weekdays = [
      l10n.commonWeekdayMon,
      l10n.commonWeekdayTue,
      l10n.commonWeekdayWed,
      l10n.commonWeekdayThu,
      l10n.commonWeekdayFri,
      l10n.commonWeekdaySat,
      l10n.commonWeekdaySun,
    ];

    double maxY = dailyCounts.reduce((a, b) => a > b ? a : b).toDouble();
    if (maxY < 5) maxY = 5;

    return BarChart(
      BarChartData(
        alignment: BarChartAlignment.spaceAround,
        maxY: maxY + 1,
        barTouchData: const BarTouchData(enabled: false),
        titlesData: FlTitlesData(
          bottomTitles: AxisTitles(
            sideTitles: SideTitles(
              showTitles: true,
              getTitlesWidget: (double value, TitleMeta meta) {
                final int dayIndex = value.toInt();
                // Only highlight the day if we are actually viewing the current week
                final bool isToday =
                    isCurrentRealWeek && (dayIndex == todayIndex);

                return Padding(
                  padding: const EdgeInsets.only(top: 8.0),
                  child: Text(
                    weekdays[dayIndex],
                    style: TextStyle(
                      fontSize: 12,
                      fontWeight: isToday ? FontWeight.bold : FontWeight.normal,
                      color: isToday
                          ? AppTheme.taskColor
                          : Theme.of(context).colorScheme.onSurfaceVariant,
                    ),
                  ),
                );
              },
            ),
          ),
          leftTitles: const AxisTitles(),
          topTitles: const AxisTitles(),
          rightTitles: const AxisTitles(),
        ),
        gridData: const FlGridData(show: false),
        borderData: FlBorderData(show: false),
        barGroups: List.generate(7, (index) {
          final bool isToday = isCurrentRealWeek && (index == todayIndex);
          return BarChartGroupData(
            x: index,
            barRods: [
              BarChartRodData(
                toY: dailyCounts[index].toDouble(),
                color: isToday
                    ? AppTheme.taskColor
                    : AppTheme.taskColor.withValues(
                        alpha: AppTheme.opacityMedium,
                      ),
                width: 16,
                borderRadius: BorderRadius.circular(4),
                backDrawRodData: BackgroundBarChartRodData(
                  show: true,
                  toY: maxY + 1,
                  color: Colors.grey.withValues(alpha: 0.1),
                ),
              ),
            ],
          );
        }),
      ),
    );
  }

  // Helper to jump weeks
  void _changeWeek(int days) {
    setState(() {
      _focusedDate = _focusedDate.add(Duration(days: days));
    });
  }

  List<List<int>> _getVelocityForWeek(
    List<Task> tasks,
    DateTime startOfWeek,
    DateTime endOfWeek,
  ) {
    final List<List<int>> dailyCounts = List.generate(7, (_) => [0, 0]);

    for (var task in tasks) {
      if (task.createdAt.isAfter(
            startOfWeek.subtract(const Duration(seconds: 1)),
          ) &&
          task.createdAt.isBefore(endOfWeek)) {
        dailyCounts[task.createdAt.weekday - 1][0]++;
      }
      if (task.isCompleted && task.completedAt != null) {
        if (task.completedAt!.isAfter(
              startOfWeek.subtract(const Duration(seconds: 1)),
            ) &&
            task.completedAt!.isBefore(endOfWeek)) {
          dailyCounts[task.completedAt!.weekday - 1][1]++;
        }
      }
    }
    return dailyCounts;
  }

  List<int> _getTasksCompletedForWeek(
    List<Task> tasks,
    DateTime startOfWeek,
    DateTime endOfWeek,
  ) {
    final List<int> dailyCounts = List.filled(7, 0);

    for (var task in tasks) {
      if (task.isCompleted && task.completedAt != null) {
        if (task.completedAt!.isAfter(
              startOfWeek.subtract(const Duration(seconds: 1)),
            ) &&
            task.completedAt!.isBefore(endOfWeek)) {
          dailyCounts[task.completedAt!.weekday - 1]++;
        }
      }
    }
    return dailyCounts;
  }

  // Category data remains "All Time" to show overall distribution
  Map<String, int> _getCategoryData(List<Task> tasks) {
    final Map<String, int> categoryCounts = {};
    for (var task in tasks) {
      final String cat = task.category.isEmpty
          ? AppLocalizations.of(context)!.taskStatsCategoryUncategorized
          : task.category;
      categoryCounts[cat] = (categoryCounts[cat] ?? 0) + 1;
    }
    return categoryCounts;
  }

  Widget _buildCategoryBreakdownCard(BuildContext context, List<Task> tasks) {
    final categoryData = _getCategoryData(tasks);
    final l10n = AppLocalizations.of(context)!;
    if (categoryData.isEmpty) {
      return Center(child: Text(l10n.taskStatsCategoryUnused));
    }

    final totalTasks = categoryData.values.fold<int>(
      0,
      (sum, value) => sum + value,
    );
    final entries = categoryData.entries.toList()
      ..sort((a, b) => b.value.compareTo(a.value));
    final colors = [
      AppTheme.taskColor,
      AppTheme.errorColor,
      AppTheme.successColor,
      AppTheme.warningColor,
      AppTheme.habitColor,
      Theme.of(context).colorScheme.primary,
      Colors.teal,
      Colors.deepOrange,
    ];

    Color colorForIndex(int index) => colors[index % colors.length];

    // Removed Card and Container background styling to fit with the rest of the screen elements
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          l10n.taskStatsDistribution,
          style: const TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
        ),
        const SizedBox(height: 4),
        Text(
          l10n.taskStatsDistributionSubtitle,
          style: TextStyle(
            fontSize: 12,
            color: Theme.of(context).colorScheme.onSurfaceVariant,
          ),
        ),
        const SizedBox(height: 18),
        ClipRRect(
          borderRadius: BorderRadius.circular(999),
          child: Container(
            height: 16,
            color: Theme.of(context).colorScheme.surfaceContainerHighest,
            child: Row(
              children: entries.asMap().entries.map((entry) {
                final index = entry.key;
                final category = entry.value;
                return Expanded(
                  flex: category.value,
                  child: Container(color: colorForIndex(index)),
                );
              }).toList(),
            ),
          ),
        ),
        const SizedBox(height: 18),
        ...entries.asMap().entries.map((entry) {
          final index = entry.key;
          final category = entry.value;
          final color = colorForIndex(index);
          final percent = ((category.value / totalTasks) * 100).round();

          return Padding(
            padding: const EdgeInsets.only(bottom: 12),
            child: Row(
              children: [
                Container(
                  width: 14,
                  height: 14,
                  decoration: BoxDecoration(
                    color: color,
                    shape: BoxShape.circle,
                  ),
                ),
                const SizedBox(width: 10),
                Expanded(
                  child: Text(
                    category.key,
                    style: const TextStyle(
                      fontSize: 14,
                      fontWeight: FontWeight.w600,
                    ),
                    overflow: TextOverflow.ellipsis,
                  ),
                ),
                Text(
                  l10n.nTasksCount(category.value),
                  style: const TextStyle(
                    fontSize: 14,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                const SizedBox(width: 12),
                Text(
                  '$percent%',
                  style: TextStyle(
                    fontSize: 12,
                    color: Theme.of(context).colorScheme.onSurfaceVariant,
                  ),
                ),
              ],
            ),
          );
        }),
      ],
    );
  }
}

File: app/src/main/java/io/github/benji377/timety/ui/screens/task/TaskStatsScreen.kt
package io.github.benji377.timety.ui.screens.task

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.WarningAmber
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import io.github.benji377.timety.R
import io.github.benji377.timety.data.model.focus.FocusSessionEntity
import io.github.benji377.timety.data.model.task.TaskWithSubtasks
import io.github.benji377.timety.ui.components.common.WeekNavigator
import io.github.benji377.timety.ui.components.stats.LegendDot
import io.github.benji377.timety.ui.components.stats.SectionHeader
import io.github.benji377.timety.ui.theme.ChartDeepOrange
import io.github.benji377.timety.ui.theme.ChartTeal
import io.github.benji377.timety.ui.theme.ErrorColor
import io.github.benji377.timety.ui.theme.HabitColor
import io.github.benji377.timety.ui.theme.SuccessColor
import io.github.benji377.timety.ui.theme.TaskColor
import io.github.benji377.timety.ui.theme.WarningAccent
import io.github.benji377.timety.ui.theme.WarningColor
import io.github.benji377.timety.ui.utils.AppUtils
import io.github.benji377.timety.ui.utils.quantityString
import io.github.benji377.timety.ui.viewmodel.AppViewModelProvider
import io.github.benji377.timety.ui.viewmodel.FocusViewModel
import io.github.benji377.timety.ui.viewmodel.RecurringTaskViewModel
import io.github.benji377.timety.ui.viewmodel.TaskViewModel
import io.github.benji377.timety.ui.viewmodel.activityScopedViewModel
import io.github.benji377.timety.util.datetime.AppDateFormatUtils
import io.github.benji377.timety.util.datetime.AppDateUtils
import io.github.benji377.timety.util.stats.CalibrationBucket
import io.github.benji377.timety.util.stats.CalibrationInsight
import io.github.benji377.timety.util.stats.EstimationCalibrator
import io.github.benji377.timety.util.stats.StatsUtils
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import kotlin.math.roundToInt


/**
 * Shows weekly task velocity and productivity charts plus an all-time category breakdown for the
 * selected week.
 */
@Composable
fun TaskStatsScreen(
    taskViewModel: TaskViewModel = viewModel(factory = AppViewModelProvider.Factory),
    recurringViewModel: RecurringTaskViewModel = viewModel(factory = AppViewModelProvider.Factory),
    focusViewModel: FocusViewModel = activityScopedViewModel(),
) {
    val tasks by taskViewModel.allTasks.collectAsState()
    val recurringItems by recurringViewModel.allRecurringTasks.collectAsState()
    val sessions by focusViewModel.allSessions.collectAsState()
    var focusedDate by remember { mutableStateOf(LocalDate.now()) }

    // Recurring tasks feed the charts too: templates count as created once, each logged
    // occurrence as one completion.
    val recurringCreated = remember(recurringItems) { recurringItems.map { it.task.createdAt } }
    val recurringCompleted = remember(recurringItems) {
        recurringItems.flatMap { item -> item.occurrences.map { it.completedAt } }
    }

    val zone = ZoneId.systemDefault()
    val startOfWeek = AppDateUtils.startOfWeekMonday(focusedDate)
    val endOfWeekLocal = startOfWeek.plusDays(6)
    val startOfWeekInstant = startOfWeek.atStartOfDay(zone).toInstant()
    val endOfWeekInstant = endOfWeekLocal.atTime(23, 59, 59).atZone(zone).toInstant()

    val isCurrentRealWeek =
        AppDateUtils.isWithinInclusive(LocalDate.now(), startOfWeek, endOfWeekLocal)

    if (tasks.isEmpty() && recurringItems.isEmpty()) {
        Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
            Text(stringResource(R.string.taskStatsEmpty))
        }
    } else {
        LazyColumn(
            modifier = Modifier
                .fillMaxSize()
                .padding(16.dp)
        ) {
            item {
                WeekNavigator(
                    focusedDate = focusedDate,
                    onShiftWeek = { days -> focusedDate = focusedDate.plusDays(days.toLong()) }
                )
            }
            item { Spacer(modifier = Modifier.height(16.dp)) }

            // Task velocity chart.
            item {
                SectionHeader(stringResource(R.string.taskStatsVelocity))
                Spacer(modifier = Modifier.height(4.dp))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    LegendDot(WarningColor)
                    Spacer(modifier = Modifier.width(4.dp))
                    Text(
                        stringResource(R.string.taskStatsCreated),
                        fontSize = 12.sp,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                    Spacer(modifier = Modifier.width(16.dp))
                    LegendDot(SuccessColor)
                    Spacer(modifier = Modifier.width(4.dp))
                    Text(
                        stringResource(R.string.taskStatsCompleted),
                        fontSize = 12.sp,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
                Spacer(modifier = Modifier.height(16.dp))
                Box(modifier = Modifier.height(200.dp)) {
                    VelocityChart(
                        tasks,
                        recurringCreated,
                        recurringCompleted,
                        startOfWeekInstant,
                        endOfWeekInstant,
                        isCurrentRealWeek,
                        zone
                    )
                }
            }
            item { Spacer(modifier = Modifier.height(40.dp)) }

            // Productivity bar chart.
            item {
                SectionHeader(
                    stringResource(R.string.taskStatsProductivity),
                    stringResource(R.string.taskStatsCompletedDaily)
                )
                Spacer(modifier = Modifier.height(16.dp))
                Box(modifier = Modifier.height(200.dp)) {
                    ProductivityChart(
                        tasks,
                        recurringCompleted,
                        startOfWeekInstant,
                        endOfWeekInstant,
                        isCurrentRealWeek,
                        zone
                    )
                }
            }
            item { Spacer(modifier = Modifier.height(40.dp)) }

            // Category breakdown, all time.
            item { CategoryBreakdownCard(tasks, recurringItems.map { it.task.category }) }
            item { Spacer(modifier = Modifier.height(40.dp)) }

            // Estimation calibration, all time.
            item { EstimationCalibrationCard(tasks, sessions) }
            item { Spacer(modifier = Modifier.height(40.dp)) }
        }
    }
}

@Composable
private fun VelocityChart(
    tasks: List<TaskWithSubtasks>,
    recurringCreated: List<Instant>,
    recurringCompleted: List<Instant>,
    startOfWeek: Instant,
    endOfWeek: Instant,
    isCurrentRealWeek: Boolean,
    zone: ZoneId
) {
    val dailyCreated = IntArray(7)
    val dailyCompleted = IntArray(7)
    fun IntArray.countInWeek(instant: Instant) {
        if (!instant.isBefore(startOfWeek) && instant.isBefore(endOfWeek)) {
            this[instant.atZone(zone).dayOfWeek.value - 1]++
        }
    }

    tasks.forEach {
        dailyCreated.countInWeek(it.task.createdAt)
        if (it.task.isCompleted) {
            it.task.completedAt?.let { completedAt -> dailyCompleted.countInWeek(completedAt) }
        }
    }
    recurringCreated.forEach { dailyCreated.countInWeek(it) }
    recurringCompleted.forEach { dailyCompleted.countInWeek(it) }

    val maxY = StatsUtils.maxValue((dailyCreated.toList() + dailyCompleted.toList()), minimum = 5.0)

    SimpleBarChart(
        data1 = dailyCreated,
        color1 = WarningColor,
        data2 = dailyCompleted,
        color2 = SuccessColor,
        maxVal = maxY,
        isCurrentRealWeek = isCurrentRealWeek,
        zone = zone
    )
}

@Composable
private fun ProductivityChart(
    tasks: List<TaskWithSubtasks>,
    recurringCompleted: List<Instant>,
    startOfWeek: Instant,
    endOfWeek: Instant,
    isCurrentRealWeek: Boolean,
    zone: ZoneId
) {
    val dailyCompleted = IntArray(7)
    val completions = tasks.mapNotNull { if (it.task.isCompleted) it.task.completedAt else null } +
            recurringCompleted

    completions.forEach { completedAt ->
        if (!completedAt.isBefore(startOfWeek) && completedAt.isBefore(endOfWeek)) {
            dailyCompleted[completedAt.atZone(zone).dayOfWeek.value - 1]++
        }
    }

    val maxY = StatsUtils.maxValue(dailyCompleted.toList(), minimum = 5.0)

    SimpleBarChart(
        data1 = dailyCompleted,
        color1 = TaskColor,
        data2 = null,
        color2 = Color.Transparent,
        maxVal = maxY,
        isCurrentRealWeek = isCurrentRealWeek,
        zone = zone
    )
}

@Composable
private fun SimpleBarChart(
    data1: IntArray,
    color1: Color,
    data2: IntArray?,
    color2: Color,
    maxVal: Double,
    isCurrentRealWeek: Boolean,
    zone: ZoneId
) {
    val days = listOf(
        stringResource(R.string.commonWeekdayMon),
        stringResource(R.string.commonWeekdayTue),
        stringResource(R.string.commonWeekdayWed),
        stringResource(R.string.commonWeekdayThu),
        stringResource(R.string.commonWeekdayFri),
        stringResource(R.string.commonWeekdaySat),
        stringResource(R.string.commonWeekdaySun)
    )
    val todayIndex = if (isCurrentRealWeek) Instant.now().atZone(zone).dayOfWeek.value - 1 else -1
    // Add one unit of headroom above the tallest bar so it doesn't touch the top edge.
    val denom = (maxVal + 1).toFloat()

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(200.dp),
        horizontalArrangement = Arrangement.SpaceEvenly,
        verticalAlignment = Alignment.Bottom
    ) {
        for (i in 0 until 7) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Bottom,
                modifier = Modifier
                    .weight(1f)
                    .fillMaxHeight()
            ) {
                Row(
                    modifier = Modifier.weight(1f),
                    horizontalArrangement = Arrangement.Center,
                    verticalAlignment = Alignment.Bottom
                ) {
                    val h1 = (data1[i].toFloat() / denom).coerceIn(0f, 1f)
                    Box(
                        modifier = Modifier
                            .fillMaxHeight(h1)
                            .width(12.dp)
                            .clip(RoundedCornerShape(topStart = 4.dp, topEnd = 4.dp))
                            .background(color1)
                    )

                    if (data2 != null) {
                        Spacer(modifier = Modifier.width(2.dp))
                        val h2 = (data2[i].toFloat() / denom).coerceIn(0f, 1f)
                        Box(
                            modifier = Modifier
                                .fillMaxHeight(h2)
                                .width(12.dp)
                                .clip(RoundedCornerShape(topStart = 4.dp, topEnd = 4.dp))
                                .background(color2)
                        )
                    }
                }
                Spacer(modifier = Modifier.height(8.dp))
                val isToday = i == todayIndex
                Text(
                    text = days[i],
                    style = MaterialTheme.typography.bodySmall.copy(fontWeight = if (isToday) FontWeight.Bold else FontWeight.Normal),
                    color = if (isToday) TaskColor else MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}

@Composable
private fun CategoryBreakdownCard(
    tasks: List<TaskWithSubtasks>,
    recurringCategories: List<String>,
) {
    val uncategorized = stringResource(R.string.taskStatsCategoryUncategorized)
    val categoryCounts = remember(tasks, recurringCategories) {
        val counts = LinkedHashMap<String, Int>()
        (tasks.map { it.task.category } + recurringCategories).forEach {
            val cat = it.ifBlank { uncategorized }
            counts[cat] = (counts[cat] ?: 0) + 1
        }
        counts
    }

    if (categoryCounts.isEmpty()) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 16.dp),
            contentAlignment = Alignment.Center
        ) {
            Text(stringResource(R.string.taskStatsCategoryUnused))
        }
        return
    }

    val totalTasks = categoryCounts.values.sum()
    val entries = categoryCounts.entries.sortedByDescending { it.value }
    val colors = listOf(
        TaskColor,
        ErrorColor,
        SuccessColor,
        WarningColor,
        HabitColor,
        MaterialTheme.colorScheme.primary,
        ChartTeal,
        ChartDeepOrange
    )

    Column {
        SectionHeader(
            stringResource(R.string.taskStatsDistribution),
            stringResource(R.string.taskStatsDistributionSubtitle)
        )
        Spacer(modifier = Modifier.height(18.dp))

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(16.dp)
                .clip(RoundedCornerShape(999.dp))
                .background(MaterialTheme.colorScheme.surfaceVariant)
        ) {
            entries.forEachIndexed { index, entry ->
                Box(
                    modifier = Modifier
                        .weight(entry.value.toFloat())
                        .fillMaxHeight()
                        .background(colors[index % colors.size])
                )
            }
        }

        Spacer(modifier = Modifier.height(18.dp))

        entries.forEachIndexed { index, entry ->
            val color = colors[index % colors.size]
            val percent = ((entry.value.toFloat() / totalTasks) * 100).roundToInt()
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 12.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Box(
                    modifier = Modifier
                        .size(14.dp)
                        .background(color, CircleShape)
                )
                Spacer(modifier = Modifier.width(10.dp))
                Text(
                    entry.key,
                    modifier = Modifier.weight(1f),
                    fontSize = 14.sp,
                    fontWeight = FontWeight.SemiBold,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
                Text(
                    quantityString(
                        R.plurals.nTasksCount,
                        entry.value,
                        zeroRes = R.string.nTasksCountZero,
                        entry.value
                    ),
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Bold
                )
                Spacer(modifier = Modifier.width(12.dp))
                Text(
                    "$percent%",
                    fontSize = 12.sp,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}

@Composable
private fun EstimationCalibrationCard(
    tasks: List<TaskWithSubtasks>,
    sessions: List<FocusSessionEntity>,
) {
    val buckets = remember(tasks, sessions) {
        EstimationCalibrator.buckets(
            tasks.filter { it.task.isCompleted }.map { it.task },
            sessions
        )
    }

    Column {
        SectionHeader(
            stringResource(R.string.taskStatsCalibrationTitle),
            stringResource(R.string.taskStatsCalibrationSubtitle)
        )
        Spacer(modifier = Modifier.height(18.dp))

        if (buckets.isEmpty()) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 16.dp),
                contentAlignment = Alignment.Center
            ) {
                Text(stringResource(R.string.taskStatsCalibrationEmpty))
            }
            return
        }

        buckets.forEach { bucket ->
            CalibrationBucketRow(bucket)
            Spacer(modifier = Modifier.height(12.dp))
        }

        val insight = remember(buckets) { EstimationCalibrator.insight(buckets) }
        if (insight != null) {
            Spacer(modifier = Modifier.height(4.dp))
            Row(verticalAlignment = Alignment.Top) {
                Icon(
                    Icons.Filled.WarningAmber,
                    contentDescription = null,
                    tint = WarningAccent,
                    modifier = Modifier.size(18.dp)
                )
                Spacer(modifier = Modifier.width(8.dp))
                Text(
                    calibrationInsightText(insight),
                    fontSize = 13.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = WarningAccent,
                )
            }
        }
    }
}

@Composable
private fun CalibrationBucketRow(bucket: CalibrationBucket) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Box(
            modifier = Modifier
                .size(32.dp)
                .clip(CircleShape)
                .background(TaskColor.copy(alpha = 0.12f)),
            contentAlignment = Alignment.Center
        ) {
            Text(
                AppUtils.getSizeEmoji(bucket.size),
                fontSize = 12.sp,
                fontWeight = FontWeight.Bold,
                color = TaskColor,
            )
        }
        Spacer(modifier = Modifier.width(12.dp))
        Text(
            "~${AppDateFormatUtils.formatMinutesCompact(bucket.avgMinutes)}",
            modifier = Modifier.weight(1f),
            fontSize = 14.sp,
            fontWeight = FontWeight.Bold
        )
        Text(
            quantityString(
                R.plurals.nTasksCount,
                bucket.sampleCount,
                zeroRes = R.string.nTasksCountZero,
                bucket.sampleCount
            ),
            fontSize = 13.sp,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
private fun calibrationInsightText(insight: CalibrationInsight): String = when (insight) {
    is CalibrationInsight.OrderingViolation -> stringResource(
        R.string.taskStatsCalibrationOrderingViolation,
        insight.largerSize.name.replace("_", ""),
        insight.smallerSize.name.replace("_", ""),
    )

    is CalibrationInsight.HighSpread -> stringResource(
        R.string.taskStatsCalibrationHighSpread,
        insight.size.name.replace("_", ""),
    )
}
